roaring/container_stash.go

```diff
@@ -0,0 +1,263 @@
+}
+
+// NewContainerArray returns an array using the provided set of values. It's
+// okay if the slice is nil; that's a length of zero.
+func NewContainerArray(set []uint16) *Container {
+	c := &Container{typ: containerArray, n: int32(len(set))}
+	c.setArray(set)
+	return c
+}
+
+// NewContainerRun creates a new run array using a provided (possibly nil)
+// slice of intervals.
+func NewContainerRun(set []interval16) *Container {
+	c := &Container{typ: containerRun}
+	c.setRuns(set)
+	for _, run := range set {
+		c.n += int32(run.last-run.start) + 1
+	}
+	return c
+}
+
+// Mapped returns the internal mapped field, which indicates whether the
+// slice's backing store is believed to be associated with unwriteable
+// mmapped space.
+func (c *Container) Mapped() bool {
+	return c.mapped
+}
+
+// N returns the internal n field.
+func (c *Container) N() int32 {
+	return c.n
+}
+
+// array yields the data viewed as a slice of uint16 values.
+func (c *Container) array() []uint16 {
+	if roaringParanoia {
+		if c.typ != containerArray {
+			panic("attempt to read non-array's array")
+		}
+	}
+	return *(*[]uint16)(unsafe.Pointer(&reflect.SliceHeader{Data: uintptr(unsafe.Pointer(c.pointer)), Len: int(c.len), Cap: int(c.cap)}))
+}
+
+// setArray stores a set of uint16s as data.
+func (c *Container) setArray(array []uint16) {
+	if roaringParanoia {
+		if c.typ != containerArray {
+			panic("attempt to write non-array's array")
+		}
+	}
+	// no array: start with our default 5-value array
+	if array == nil {
+		c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(&c.data[0])), 0, stashedArraySize
+		return
+	}
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&array))
+	if h.Data == uintptr(unsafe.Pointer(c.pointer)) {
+		// nothing to do but update length
+		c.len = int32(h.Len)
+		return
+	}
+	// array we can fit in data store:
+	if len(array) <= stashedArraySize {
+		copy(c.data[:stashedArraySize], array)
+		c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(&c.data[0])), int32(len(array)), stashedArraySize
+		c.mapped = false // this is no longer using a hypothetical mmapped input array
+		return
+	}
+	c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(h.Data)), int32(h.Len), int32(h.Cap)
+	runtime.KeepAlive(&array)
+}
+
+// bitmap yields the data viewed as a slice of uint64s holding bits.
+func (c *Container) bitmap() []uint64 {
+	if roaringParanoia {
+		if c.typ != containerBitmap {
+			panic("attempt to read non-bitmap's bitmap")
+		}
+	}
+	return *(*[]uint64)(unsafe.Pointer(&reflect.SliceHeader{Data: uintptr(unsafe.Pointer(c.pointer)), Len: int(c.len), Cap: int(c.cap)}))
+}
+
+// setBitmap stores a set of uint64s as data.
+func (c *Container) setBitmap(bitmap []uint64) {
+	if roaringParanoia {
+		if c.typ != containerBitmap {
+			panic("attempt to write non-bitmap's bitmap")
+		}
+	}
+	h := (*reflect.SliceHeader)(unsafe.Pointer(&bitmap))
+	c.pointer, c.len, c.cap = (*uint16)(unsafe.Pointer(h.Data)), int32(h.Len), int32(h.Cap)
+	runtime.KeepAlive(&bitmap)
```

Comment:
I don't think I understand the use of KeepAlive—is the idea that as soon as you `unsafe.Pointer(&bitmap)` the bitmap could be GCd?

Reply:
Basically, if there's no references to `bitmap` itself after line 147, the GC is in principle allowed to discard the bitmap slice header, because lifetime analysis lets it realize that the local variable isn't *actually* still in use. The reflect.SliceHeader doesn't count as having a live pointer, so it's possible that, if this was the only reference to the slice's backing store (the caller's copy could also be getting nixed by lifetyme analysis), the backing store gets dropped. It's not very likely, I think, but also runtime.KeepAlive is basically free; it effectively serves as a compiler hint.